### ISSUE TITLE
Create a Dynamic Form

### ISSUE THREAD

**Issue by mattbatman** [CivicCode/uReport-client#4]
We need to create a dynamic form component that can consume the response data from the queries (as described in #3) and populate itself with the correct fields. This will also involve making components for each individual field.

## Field Components
- [ ] Location selection component
- [ ] Input component
- [ ] Text area component
- [ ] Dropdown selection component
- [ ] Checkbox component
- [ ] Media upload component
- [ ] Captcha component

### On Load
- [ ] Make a call to the `serviceDefinition` if `metadata` was true in the call for `serviceList`
- [ ] I guess the bulk of the questions are included with the `serviceList` call?

### On Submit
- [ ] Make a call to the API to post

## Functionality
- [ ] Dynamically render fields only as needed

**Comment by mattbatman** [comment]
I think the two most time intensive things left to do for the form before Tuesday night will be:
1. create/handle the Google map location selection
2. do something to for CAPTCHA (I think they use [this)[https://www.google.com/recaptcha/intro/])

The other significant items remaining, I think, are just:
1. wire in the fields from our `serviceDefinition` once the return type is fixed
2. handle checkbox or multi-value selection fields
3. make a `POST` on form submission

**Comment by matt-a-t** [comment]
I was testing tonight, and I could not get multi-value selection fields to recognize more than one value in the api. I added a switch in to handle multivalue types in the same way as singlevalue types in the dropdown.

I also made some changes to the file submission process. When a file is uploaded it is sent to a Cloudinary account and the url of the image is returned. This will be submitted in the POST request as the media_url parameter, and is saved in the state of the Form that is currently returned on media submit.

This likely still needs some styling and we still need to resolve the POST request.

**Comment by mattbatman** [comment]
@dpalazz I saw your comment in the routing ticket. I'd say take your pick of:

- getting select (and multi-select) form fields to work
  - I was leaning towards installing [react-select](http://jedwatson.github.io/react-select/)
- handling address/location input
  - they used a Google map picker thing
- handling date input
- creating a form `POST` on submit
- general form styling

**Comment by matt-a-t** [comment]
Multi-select form fields are working as well as they can with the API. The documentation says that multiple values can be submitted as attribute[code]=value&attribute[code]=value, but using Postman I was only able to get the last entry recognized by their server. So I have it handled in the same way as the singlevalue type in a dropdown menu.

**Comment by dpalazz** [comment]
When you say getting those select fields to work - what do you mean exactly?

In the meantime I will look into the handling address/location input.

**Comment by dpalazz** [comment]
Is anyone working on the app right now?

**Comment by mattbatman** [comment]
I just started to.

**Comment by dpalazz** [comment]
I was able to display the map using react-location-picker. Now we just need
to translate that into the form submission. Are you working on that part
yet (handlesubmit)?

**Comment by mattbatman** [comment]
Not directly. I just pushed a commit to try to fix how the form tracks state for each field.

**Comment by dpalazz** [comment]
Cool, just pushed what I had for location picker.

**Comment by matt-a-t** [comment]
If we can get state on the Form component set I will work out the post query tomorrow

**Comment by mattbatman** [comment]
> The location picker updates an <h1> element dynamically with an address
> when the user moves the red pin. Could we use that to feed into the form?

Yep. Great job with this. I think we need either 1) the address string or 2) latitude and longitude coordinates. I'd imagine the address string is fine, and it seems slightly easier than the coordinates.

I'm guessing the correct way to do this is to pass a handler from the parent `Form` component as a prop to the `MapPicker` component.

**Comment by dpalazz** [comment]
Okay, I'm going to push some style changes (let me know if you don't like
the direction I took) then I'll take a look at this. Thanks!

**Comment by mattbatman** [comment]
@dpalazz I have the location picker wired into the form. I'll wait to push it so you don't get a merge conflict.

**Comment by dpalazz** [comment]
Oops - sorry, taking a long time on this design. Longer than I should. I'm
almost done and I'll push it after I pull down ur changes.

**Comment by matt-a-t** [comment]
Thanks for getting all this done tonight everybody. It looks like we should be in good shape to get things wired up tomorrow

**Comment by dpalazz** [comment]
Hittin' the sack for the night. Thanks ya'll!
